### PR TITLE
chore(exoflex): update `react`, `react-native` & `react-native-paper` version

### PR DESCRIPTION
**Pull request by StefanusChristian**
### Changes:

**Package & Example app changes:**
- Upgrade `react`, `react-native`, `react-native-web`, `react-native-paper`, `react-native-collapsible` version.
- Change `react-native-calendars` from Yoan repository into latest version. (changes from @JonathanRadotski  https://github.com/StefanusChristian/infra/pull/1/files)
- Add new expo project in `example` directory. Since the previous one was still in SDK40, it is easier to create a new one instead of updating it to SDK47.
- Add `patch-package` to both `exoflex` (patch `react-native-collapsible`) & `exoflex/new-example` (patch `react-navigation-drawer`). 
   - `react-native-collapsible`: Adding the fix that has been merged to the package but not released yet. [commit](https://github.com/oblador/react-native-collapsible/commit/4e0d4ce7b337f702d8986b39d0c71e90954ea0bf)
   - `react-navigation-drawer`: This package is deprecated now and have been included into `react-navigation` instead. But I think it was introduced in version `5.x.x` ([docs](https://reactnavigation.org/docs/5.x/drawer-based-navigation))  and we still use `4.x.x` ([docs](https://reactnavigation.org/docs/4.x/drawer-based-navigation)). So for a quick fix, I follow this guide for the error: https://github.com/react-navigation/react-navigation/issues/9665#issuecomment-880711983 

**Adjusted Component:**
- `Accordion`: Add `useNativeRiver` props to `useAnimation()`
- `Badge`: Add `textStyle` typing
- `Button`:
     - `ButtonOpacity`: 
         - Update ref as `RefObject<TouchableOpacity>` since `ButtonProp` types based on `react-native-paper` types and the ref was `RefObject<View>` (from Surface component).  
         - Adjust usage of `accessibilityTraits` & `accessibilityComponentType` into `accessibilityStates`. ([Reference](https://reactnative.dev/blog/2018/08/13/react-native-accessibility-updates#solution-one-combining-accessibilitycomponenttype-android-and-accessibilitytraits-ios))
         - Adjust `Icon` props naming from `color` to `iconColor`.
    - `ButtonRipple`:
        - Adjust usage of `accessibilityTraits` & `accessibilityComponentType` into `accessibilityStates`.
        - Adjust `Icon` props naming from `color` to `iconColor`.
- `Calendar`: Adjust `CalendarProps` to follow latest version of the package.
- `Chip`: Adjust `Icon` props naming from `color` to `iconColor`.
- `Collapsible`: Add `useNativeRiver` props to `useAnimation()`
- `DateTimePicker`: Adjust `DateTimePicker.web` type from `DateObject` into `DateData`.
- `Menu`: Adjust `PaperMenu.Item` props naming from `icon` to `leadingIcon`.
- `SegmentedControl`: Add `StyleProp<ViewStyle>` in both `Indicator` & `SegmentedControl`.
- `Switch`:  Add `StyleProp<ViewStyle>` types.
- `Toast`: Adjust `Icon` props naming from `color` to `iconColor`.

**Contants, Helper & Types Changes:**
- `Theme`: Add 2 new color key as part of the Theme: `notification` & `tooltip`. 
- `getPaperTheme`: Adjust import naming from `Theme` into `MD2Theme` and adjust the props needed.
- `Colors` types: Add new key `tooltip` as Colors type.
- Custom `react-native-color-slider` types: Change `number[]` into `Array<number>`

### **How to test:**
- Move to `packages/exoflex` and run `yarn install` to install dependencies.
- Move to `packages/exoflex/example` and also run `yarn install`. Once it's done, run `yarn start` to test the component.

### **Notes:**
Based on testing, the only component that has problem is `ButtonRipple`. The button didn't work properly on iOS simulator, but animation run properly on Android (cc: @JonathanRadotski @Udayawibawamukti).

### CODE REVIEW

File: packages/exoflex/package.json
Comment (ikusa) [inline packages/exoflex/package.json:40]: On reverting this to the newest feat, have we test these change support what Yoan originally patch out?

File: packages/exoflex/example/metro.config.js
Comment (ikusa) [inline packages/exoflex/example/metro.config.js:46]: let's remove these

File: packages/exoflex/example/patches/react-navigation-drawer+2.7.2.patch
Comment (ikusa) [inline packages/exoflex/example/patches/react-navigation-drawer+2.7.2.patch:4]: After checking the issue, from my understanding it seems this was caused by we using react-navigation 4. It's not required to do this now but I think it's good to also upgrade it to react-navigation 5/6.

File: packages/exoflex/src/components/Accordion.tsx
Comment (ikusa) [inline packages/exoflex/src/components/Accordion.tsx:155]: I've checked out these changes and it seems now <View> will throw error when we use more than one child. Is this supposed to work like that? or is it possible we messed up with something? React native [doc](https://reactnative.dev/docs/view#props) say `View is designed to be nested inside other views and can have 0 to many children of any type.`

File: packages/exoflex/tsconfig.json
Comment (ikusa) [inline packages/exoflex/tsconfig.json:15]: let's revert this since we already wipe the new-example folder

**Review by ikusa — APPROVED** [review]
Thanks for upgrading all of these. The changes looks good but I've several suggestion and question. We also might need @darcien help to review these too, since I'm not super familiar with exoflex

**Comment by ikusa** [comment]
@StefanusChristian Also do we have the example of buttonRipple not working properly on the IOS? curious what's the actual result of these bug

File: packages/exoflex/src/components/Button/ButtonOpacity.tsx
Comment (darcien) [inline packages/exoflex/src/components/Button/ButtonOpacity.tsx:42]: I'm not sure if passing `ref` directly will work or not.
This will need a bit of testing and see if this needs use forwardRef instead.

For now I think this is OK, just make sure to not release this before testing whether the ref works or not.

File: packages/exoflex/src/components/Badge.tsx
Comment (darcien) [inline packages/exoflex/src/components/Badge.tsx:60]: Looks a bit weird, what was the original type for `testID`?

File: packages/exoflex/src/components/Accordion.tsx
Comment (darcien) [reply packages/exoflex/src/components/Accordion.tsx:155]: Might be something from the collapsible package we use which attach a gesture handler to the children, requiring specifically only 1 child. Need investigation as I don't remember the detail here.

File: packages/exoflex/example/webpack.config.js
Comment (darcien) [inline packages/exoflex/example/webpack.config.js:1]: Just making sure here since we deleted the webpack config file from the example.

Does changes to exoflex sources reflect in the example project when running?

**Comment by darcien** [comment]
Great work @StefanusChristian! We might need to update the repo settings before merging your changes because right now it's blocked by a dead circleci check.

File: packages/exoflex/package.json
Comment (StefanusChristian) [reply packages/exoflex/package.json:40]: I think it's already supported on the newest version, but let me include @JonathanRadotski to the discussion. What do you think?

File: packages/exoflex/example/metro.config.js
Comment (StefanusChristian) [reply packages/exoflex/example/metro.config.js:46]: My bad, forgot about this. Will do!

File: packages/exoflex/example/patches/react-navigation-drawer+2.7.2.patch
Comment (StefanusChristian) [reply packages/exoflex/example/patches/react-navigation-drawer+2.7.2.patch:4]: Got it, will try and report later.

File: packages/exoflex/tsconfig.json
Comment (StefanusChristian) [reply packages/exoflex/tsconfig.json:15]: Will do

File: packages/exoflex/example/tsconfig.json
Comment (StefanusChristian) [reply packages/exoflex/example/tsconfig.json:4]: Nope, I think I comment it by accident. Will revert

File: packages/exoflex/src/components/Button/ButtonOpacity.tsx
Comment (StefanusChristian) [reply packages/exoflex/src/components/Button/ButtonOpacity.tsx:42]: Got it, will test it.

File: packages/exoflex/src/components/Badge.tsx
Comment (StefanusChristian) [reply packages/exoflex/src/components/Badge.tsx:60]: Screen Shot 2023-02-01 at 12 24 32 PM
Screen Shot 2023-02-01 at 12 25 04 PM

Since we based the props from `PaperBadge`, and it's Props type based on Animated.Text, I think it counts `testID` as `Animated.Value` here and it raises `TypeError`.

File: packages/exoflex/example/webpack.config.js
Comment (StefanusChristian) [reply packages/exoflex/example/webpack.config.js:1]: Hmm I forgot about why we remove this. 
Maybe @ikusa, did you remember why we remove this? Iirc we do this together when discussing about creating a new expo example instead of upgrading it.

But to answer your question, yes changes on exoflex sources reflect directly on the running project.

**Comment by StefanusChristian** [comment]
@ikusa @darcien 
Here's the video showing the `ButtonRipple` problem on iOS. The button `onPress` is invoked but didn't show the ripple animation at all.


https://user-images.githubusercontent.com/25154563/215960201-a3325f69-360f-48aa-ab20-292986cc07c2.mp4

File: packages/exoflex/example/patches/react-navigation-drawer+2.7.2.patch
Comment (StefanusChristian) [reply packages/exoflex/example/patches/react-navigation-drawer+2.7.2.patch:4]: Based on this: https://reactnavigation.org/docs/5.x/upgrading-from-4.x#navigation-container

I thinks it's better to upgrade on separate PR, since we still using `createAppContainer` to wrap our `RootNavigator`.

File: packages/exoflex/package.json
Comment (JonathanRadotski) [reply packages/exoflex/package.json:40]: Briefly checked and comparing added feature. I think it is supported with added feature from previous package and good to go 👍.

File: packages/exoflex/example/webpack.config.js
Comment (darcien) [reply packages/exoflex/example/webpack.config.js:1]: All good then.

For the backstory, older Expo web can't do that out of the box. So we had to customize webpack config so exoflex devs can see changes in the web example in real time.

If we don't need to do that anymore, thats great!

File: packages/exoflex/src/components/Badge.tsx
Comment (darcien) [reply packages/exoflex/src/components/Badge.tsx:60]: That is super weird, why would you want animated value in the test id.
Just to be on the safe side, could you try using String constructor instead of casting the type? Better safe on the runtime rather than sorry it crashed because of bad test id.
```suggestion
        testID={String(testID)}
```

**Comment by darcien** [comment]
Let's not block the upgrade PR on the button ripple issue. I'm keen to ship this B I G P R ASAP and address the minor issues separately.

File: packages/exoflex/src/components/Badge.tsx
Comment (StefanusChristian) [reply packages/exoflex/src/components/Badge.tsx:60]: Got it, will do!

File: packages/exoflex/src/components/Button/ButtonOpacity.tsx
Comment (StefanusChristian) [reply packages/exoflex/src/components/Button/ButtonOpacity.tsx:42]: @ikusa @darcien

About this problem, I think I'll open up a discussion in here. So the initial reason why I add this changes was because the parent exoflex `Button` component has props with type `ButtonProps` which inherit `PaperButton` types. 

`PaperButton` props type also inherit `PaperSurface` type, which has the `ref` prop with type `React.RefObject<View>`

Our exoflex Button has 2 type: `ButtonOpacity` & `ButtonRipple`, which will rendered based on `useRipple` props. 

With `ButtonRipple`, which will render `TouchableRipple` from `react-native-paper`, it seems the passed `ref` with type `<React.RefObject<View>>` from `ButtonProps` didn't raise any type error. Turns out `TouchableRipple` inherits it's prop type from RN `Pressable` and also has the `forwaredRef React.RefAttributes<View>`.

But with `ButtonOpacity`, which will render `TouchableOpacity` from `react-native`, it raise ref type error. Because the desired type was `LegacyRef<TouchableOpacity>` but supplied with `RefObject<View>`. That's why I passing the ref like that and typecast it as `RefObject<TouchableOpacity>` to fulfilled the required type.

As for using the `forwardRef`, the same problem arise when I need to type the `ref` from using `forwardRef` on Parent Button component. 

What is the best solution to handle this?

Comment (darcien) [reply packages/exoflex/src/components/Button/ButtonOpacity.tsx:42]: It's been a while since I worked on this area, so it might be a bit hard to give exact code solution.

From a glance, we should be able to resolve this type conflict by migrating away from TouchableOpacity and use Pressable instead. That way the ref is always a View ref.

The cons: we need to mimic the old TouchableOpacity using Pressable to keep the same behavior. Unless we want to change the default button behavior, and just do a breaking changes.

Comment (StefanusChristian) [reply packages/exoflex/src/components/Button/ButtonOpacity.tsx]: @ikusa @darcien 

Got it! Changes on using `Pressable` are in this commit: https://github.com/kodefox/infra/pull/713/commits/e46eaf323b3da7ff861c4b2412060b2e644ee284

I added animated opacity to make the feedback looks like `TouchableOpacity` and it seems to work as intended.


https://user-images.githubusercontent.com/25154563/216543775-798ac977-fc63-4187-a6d9-9798e40f086e.mp4

Comment (darcien) [reply packages/exoflex/src/components/Button/ButtonOpacity.tsx]: Looks great 👍 

You might want to put the animated value in a ref and memoize the fade transition functions.

Also, probably it's not a good idea to override the onPressIn/onPressOut from other props with our fade animation. Will need to extend it so we also call the onPressIn/onPressOut from the props.

Comment (StefanusChristian) [reply packages/exoflex/src/components/Button/ButtonOpacity.tsx]: Got it! Applied in here: https://github.com/kodefox/infra/pull/713/commits/4971cb9dd6c72cb62e64d49525a6fb8931ca65bb

File: packages/exoflex/src/components/Button/ButtonOpacity.tsx
Comment (darcien) [inline packages/exoflex/src/components/Button/ButtonOpacity.tsx:48]: Let's not disable this lint rule and fix the warning.

Not setting the dependency for useCallback correctly might cause subtle bugs with stale values.

Actually, let's do optimization separately now that I think about it. See my next comment.

File: packages/exoflex/src/components/Button/ButtonOpacity.tsx
Comment (darcien) [inline packages/exoflex/src/components/Button/ButtonOpacity.tsx:65]: Here, we create another arrow function inline, which is the thing we want to optimize before by moving the arrow function to useCallback.

If you want a deep dive on this, beta docs from React is a good starting point, like: https://beta.reactjs.org/reference/react/useCallback#should-you-add-usecallback-everywhere

But in the meantime, let's just keep the fix in and iterate on the optimization later.

**Review by darcien — COMMENTED** [review]
note on the memo dependency issue

File: packages/exoflex/src/components/Button/ButtonOpacity.tsx
Comment (StefanusChristian) [reply packages/exoflex/src/components/Button/ButtonOpacity.tsx:65]: Got it, will do. 
For the repo setting (remove/bypass the circleci check), do you have the authority for that @darcien? I think I don't have the permission, so I can't adjust the setting

**Comment by darcien** [comment]
>For the repo setting (remove/bypass the circleci check), do you have the authority for that @darcien? I think I don't have the permission, so I can't adjust the setting

I've removed the required CircleCI check for merging the PR.
@StefanusChristian can you try running the tests locally and see if it passes?
After that is checked, we should be good to go

**Comment by StefanusChristian** [comment]
> > For the repo setting (remove/bypass the circleci check), do you have the authority for that @darcien? I think I don't have the permission, so I can't adjust the setting
> 
> I've removed the required CircleCI check for merging the PR. In the future, we will need to migrate the CI.
> 
> @StefanusChristian can you try running the tests locally and see if it passes? After that is checked, we should be good to go

Screen Shot 2023-02-07 at 11 05 35 AM
Screen Shot 2023-02-07 at 11 07 09 AM

It seems the jest preset is error for `getBaseWebPreset`, since the `jest-preset` is removed from the latest `react-native-web` (`0.18.xx`). Reference: https://github.com/necolas/react-native-web/issues/2312

Do you guys have any idea on how to fix the `jest-preset`? @ikusa @darcien

**Comment by StefanusChristian** [comment]
@ikusa @darcien 
Web jest & test finally adjusted 🎉🎉🎉

<details>
<summary>Passed Tests</summary>

https://user-images.githubusercontent.com/25154563/218062491-74b07419-78bb-4591-a685-f1b0c7faae24.mp4

</details>

I'll add the changes for making the web test work in here:
1. Adjust the `getBaseWebPreset` (web platform jest-preset) to use module mapper based on latest rnw [docs](https://necolas.github.io/react-native-web/docs/setup/#jest) and set `textEnvironment` to `jsdom` for web.
2. Upgrade `@testing-library/react` & `react-dom` to match our current `react` version.
3. Add `testID` on some component that will turn into `data-testid` when rendered on web test. [Reference](https://stackoverflow.com/a/74819260).
4. Adjust function from `@testing-library/react`:  `wait` into `waitFor` and `fireEvent.click`. For `fireEvent` adjustment, somehow the `fireEvent.click` call doesn't trigger the mock function, so I use below form instead and it works. This might need more time to investigate but I think the adjustment is good for now
```tsx
From:

fireEvent.click(getByTestId('button'));

Into:

fireEvent(
      getByTestId('button'),
      new MouseEvent('click', {
        bubbles: true,
        cancelable: true,
      }),
    );

```

**Review by ikusa — APPROVED** [review]
Let's merge this

**Pull request merged**
